icefoganalytics/travel-authorization
# Ytgov Issue 89: Travel Auth Ability for Traveler To Update Form

Fixes https://github.com/ytgov/travel-authorization/issues/89

# Context

Relates to:

- Travel Auth Form/Approvals

# Context

**Is your feature request related to a problem? Please describe.**

- A traveler needs the ability to unsubmit a travel request to make updates or changes to reflect newest travel plans

- if the travel has not been approved allow user to just make the update and resubmit (likely need a state change so the approver doesn't approve during an edit)

- if the travel has been approved the change needs to follow the same workflow as the original process

**Describe the solution you'd like** The traveler needs the ability to reassign the travel authorization request back to themselves to make changes to dates, estimates, travel advances, travel locations, etc. The traveler should be able to select a button that says "revert to draft" if they need to update the request. Maintain the audit trail of the changes.

# Implementation

Add revert to draft button to my travel authorization "read only" view.

# Screenshots

Revert to draft action button
http://localhost:8080/my-travel-requests/6/details
![image](https://github.com/user-attachments/assets/d39f6e52-493e-4d1e-a4ff-7c2a68d9048d)

Revert to draft logs
http://localhost:8080/my-travel-requests/6/details/edit
![image](https://github.com/user-attachments/assets/5feb2fd2-6423-418d-a431-93668d06b09e)

Disabled Revert to Draft (approved travel authorization)
http://localhost:8080/my-travel-requests/6/details
![image](https://github.com/user-attachments/assets/7fba3f66-7f9d-4909-9a9f-88f51a0f9dab)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Select the "My Travel Requests" entry from the top dropdown menu.
5. Click the "+ Travel Authorization" button to create a new travel authorization request.
6. Fill out the form until you reach the "Generate Estimate" button.
7. Click the "Generate Estimate" button to generate an estimate.
8. Enter your supervisor's email address and click the "Submit to Supervisor" button.
9. Note that you are redirected to the "read-only" view of the travel authorization request.
10. Scroll to the bottom of the page and click the "Revert to Draft" button.
11. Note that you are redirected back to the "edit" view of the travel authorization request.
12. Submit the request again by clicking the "Submit to Supervisor" button.
13. Log in as the supervisor and approve the request.
14. Log in as the traveler and note that when you go back to the "read-only" view of the travel authorization request, the "Revert to Draft" button is now disabled.
15. Check that the "Revert to Draft" has an informative tooltip when disabled.
